### ISSUE TITLE
Setting the crs of naive geometries with EPSG number

### ISSUE THREAD

**Issue by jorisvandenbossche** [geopandas/geopandas#245]
Situation: you have a GeoSeries with a naive `geometry` column, and you want to set this to a certain crs (eg to convert it to another system). 
This can be done with setting the `.crs` attribute with the appropriate crs dictionary.

However, when you want to set the crs using a EPSG number, what kind of dict you have to use is not directly obvious.

Example:

```
In [44]: import geopandas
In [45]: from shapely.geometry import Point

In [57]: s = geopandas.GeoSeries([Point(4.42, 50.4), Point(4.43, 50.2)])

In [59]: s.crs = {'init': 'epsg:4326', 'no_defs': True}

In [60]: s.to_crs(epsg=31370)
Out[60]:
0     POINT (153643.9201303074 121012.188949639)
1    POINT (154373.4245113489 98766.94731544051)
dtype: object
```

The above dict is what you get if you would convert a non-naive geometry with `to_crs(epsg=..).crs`, or what you also get with fiona: 

```
In [64]: fiona.crs.from_epsg(4326)
Out[64]: {'init': 'epsg:4326', 'no_defs': True}
```

**Question**:
- We could add a `set_crs` method to `GeoSeries` and `GeoDataFrame`, which would make this more natural to do:
  
  ```
  s = s.set_crs(epsg=4326)    # instead of: s.crs = {'init': 'epsg:4326', 'no_defs': True}
  ```

Otherwise, it would be useful to add a reference to this `fiona.crs.from_epsg` function to the docs.

**Comment by micahcochran** [comment]
Is this function only for setting a GeoSeries's/GeoDataframe's crs by EPSG number?

The alternative syntax is slightly clunky, if it is documented it would be easier to find.
```python
s.crs = fiona.crs.from_epsg(4326)
```

I think that the syntax for setting `s.crs = {'proj4': 'like', 'dictionary': 'parameters'}` should be in the geopandas documentation.

**Comment by jorisvandenbossche** [comment]
Yes, setting with a EPSG number was at least my main motivation (but, of course, you could also use it for setting a crs dict, but then there is not much advantage over setting the `.crs` attribute, apart from that you can use the method in a chain)

Using `s.crs = fiona.crs.from_epsg(4326)` is indeed not that difficult, and that is the reason I asked if we would like to add a `to_crs` method, as it is somehow duplicating functionality. 
But, even if it is in the documentation, I find ``fiona.crs.from_epsg(4326)`` rather unintuitive as I wouldn't directly look for fiona to do such a thing (I would rather look at `pyproj`). And you still have to find it in the docs ..

**Comment by micahcochran** [comment]
`pyproj` doesn't support exporting proj4 projection dictionaries.  It takes a proj4 projection dictionary as input, but it won't output as a dictionary.  `pyproj`'s internal representation is a proj4 string, which it feeds that to the PROJ4 library.
```
In [22]: pyproj.Proj(init="epsg:4326", preserve_units=True).srs
Out[22]: '+init=epsg:4326 '
```

Getting back to the point,  I think `.set_crs()`  provides a function that is complimentary to the existing `.to_crs()`.  It provides a logical way to modify the crs, without having to resort to being a bit hackish.  Perhaps the syntax of `set_crs()` should be similar to `to_crs()`?  I don't see a big downside to having more than one way to set the crs.

**Comment by nickeubank** [comment]
I think `set_crs` with same syntax as `to_crs` would be great. In theory, aren't people always supposed to set important attributes directly anyway, right? And I'm with @micahcochran the syntax should be the same.

Also related to CRS manipulations: #288

**Comment by aijanai** [comment]
Hi,
I can't make the toy example that OP wrote down to work.

``` 
from shapely.geometry import Point

s = geopandas.GeoSeries([Point(4.42, 50.4), Point(4.43, 50.2)])
s.crs = {'init': 'epsg:4326', 'no_defs': True}
s.to_crs(epsg=31370)
```

yelds to:

```
---------------------------------------------------------------------------
CRSError                                  Traceback (most recent call last)
<ipython-input-153-c63352d64344> in <module>
      1 s = geopandas.GeoSeries([Point(4.42, 50.4), Point(4.43, 50.2)])
      2 s.crs = {'init': 'epsg:4326', 'no_defs': True}
----> 3 s.to_crs(epsg=31370)

/opt/conda/lib/python3.7/site-packages/geopandas/geoseries.py in to_crs(self, crs, epsg)
    302             except TypeError:
    303                 raise TypeError('Must set either crs or epsg for output.')
--> 304         proj_in = pyproj.Proj(self.crs, preserve_units=True)
    305         proj_out = pyproj.Proj(crs, preserve_units=True)
    306         if _PYPROJ2:

/opt/conda/lib/python3.7/site-packages/pyproj/proj.py in __init__(self, projparams, preserve_units, **kwargs)
    293         '116.366 39.867'
    294         """
--> 295         self.crs = CRS.from_user_input(projparams if projparams is not None else kwargs)
    296         # make sure units are meters if preserve_units is False.
    297         if not preserve_units and "foot" in self.crs.axis_info[0].unit_name:

/opt/conda/lib/python3.7/site-packages/pyproj/crs.py in from_user_input(cls, value)
    222             return cls.from_epsg(value)
    223         elif isinstance(value, dict):
--> 224             return cls(**value)
    225         elif isinstance(value, string_types):
    226             return cls.from_string(value)

/opt/conda/lib/python3.7/site-packages/pyproj/crs.py in __init__(self, projparams, **kwargs)
    144         # on case-insensitive filesystems).
    145         projstring = projstring.replace("+init=EPSG", "+init=epsg").strip()
--> 146         super(CRS, self).__init__(projstring)
    147 
    148     @classmethod

pyproj/_crs.pyx in pyproj._crs._CRS.__init__()

CRSError: Invalid projection: +init=epsg:4326 +no_defs +type=crs: (Internal Proj Error: proj_create: cannot expand +init=epsg:4326 +no_defs +type=crs)
```
What am I doing wrong? Using Python 3 with Jupyter and Conda

**Comment by jorisvandenbossche** [comment]
@aijanai can you open a new issue for that? And while doing so, can you mention the version of pyproj you are using? (`import pyproj; pyproj.__version__`)

**Comment by aijanai** [comment]
Thanks, here it is: https://github.com/geopandas/geopandas/issues/990

**Comment by snowman2** [comment]
With pyproj>=2.0 you need to remove the no_defs when using init=epsg:.. syntax. I don't know exactly why, just that it is the case.

**Comment by snowman2** [comment]
More info on nodefs not working: https://github.com/OSGeo/proj.4/issues/201

**Comment by jorisvandenbossche** [comment]
@snowman2 can you give a bit more background? (in case you understand it :-))

In any case, also with the latest pyproj (and proj 6.0.0), I don't get such error, for me `pyproj.Proj({'init': 'epsg:4326', 'no_defs': True})` works fine (it would also be a quite compatibility break to no longer accept it, even if `no_defs` is unnecessary)

**Comment by snowman2** [comment]
Based on the issue https://github.com/OSGeo/proj.4/issues/201, `no_defs` is no longer needed as it is always applied anyways.

**Comment by snowman2** [comment]
Huh, I guess I should have checked it out first.
```
>>> from pyproj import CRS
>>> cc = CRS({'init': 'epsg:4326', 'no_defs': True})
>>> cc
<CRS: +init=epsg:4326 +no_defs +type=crs>
Name: unknown
Ellipsoid:
- semi_major_metre: 6378137.00
- semi_minor_metre: 6356752.31
- inverse_flattening: 298.26
Area of Use:
- UNDEFINED
Prime Meridian:
- longitude: 0.0000
- unit_name: degree
- unit_conversion_factor: 0.01745329
Axis Info:
- Longitude[lon] (east) EPSG:9122 (degree)
- Latitude[lat] (north) EPSG:9122 (degree)
```

**Comment by snowman2** [comment]
I wonder if it was fixed here: https://github.com/pyproj4/pyproj/pull/184

**Comment by snowman2** [comment]
Nope, the boolean is not in the string.

**Comment by martinfleis** [comment]
Implemented in #1101. You can now set crs with whatever you want. All these options will work, among others:

```py
s.crs = 4326
s.crs = 'epsg:4326'
s.crs = {'init': 'epsg:4326', 'no_defs': True}
```
